Sources/SPTDataLoader/SPTDataLoaderService.m

```diff
@@ -392,10 +392,19 @@ - (void)URLSession:(NSURLSession *)session
     if (handler == nil) {
         return;
     }
-    if (handler.request.backgroundPolicy == SPTDataLoaderRequestBackgroundPolicyAlways) {
-        handler.task = [session downloadTaskWithRequest:handler.request.urlRequest];
-    } else {
-        handler.task = [session dataTaskWithRequest:handler.request.urlRequest];
+
+    //
+    // This is to avoid creating a new task on an already invalidated session
+    // when they are invalidated by `invalidateAndCancel` method.
+    // The check does not hurt even if the task is simply cancelled as there is no point
+    // in creating a new task that is not going to be used anyway.
+    //
+    if (!([error.domain isEqualToString:NSURLErrorDomain] && error.code == NSURLErrorCancelled)) {
+        if (handler.request.backgroundPolicy == SPTDataLoaderRequestBackgroundPolicyAlways) {
+            handler.task = [session downloadTaskWithRequest:handler.request.urlRequest];
+        } else {
+            handler.task = [session dataTaskWithRequest:handler.request.urlRequest];
+        }
     }
     SPTDataLoaderResponse *response = [handler completeWithError:error];
     if (response == nil && !handler.cancelled) {
```

Comment:
Seems like a reasonable change overall. I _think_ you could leverage the [existing error check](https://github.com/spotify/SPTDataLoader/blob/b0c37dc5ed258dc88ddad56fb2f3ba7fa224ace6/Sources/SPTDataLoader/SPTDataLoaderRequestTaskHandler.m#L120-L124) while keeping the same behavior by moving the task re-creation into the conditional. eg:
```suggestion

    SPTDataLoaderResponse *response = [handler completeWithError:error];
    if (response == nil && !handler.cancelled) {
        if (handler.request.backgroundPolicy == SPTDataLoaderRequestBackgroundPolicyAlways) {
            handler.task = [session downloadTaskWithRequest:handler.request.urlRequest];
        } else {
            handler.task = [session dataTaskWithRequest:handler.request.urlRequest];
        }

```